### PR TITLE
`ExecutionPlan` visitor example documentation

### LINKED ISSUE #10012: Docs for `ExecutionPlanVisitor`
### Is your feature request related to a problem or challenge?

I've had to write a couple `ExecutionPlanVisitor`s recently (see below) and when I started I initially looked for some documentation on this but wasn't able to find any.  I think it would be beneficial to new comers to see a couple examples of `ExecutionPlanVisitor` in the docs.

### Describe the solution you'd like

A section in the docs with some example implementations of `ExecutionPlanVisitor`

### Describe alternatives you've considered

_No response_

### Additional context

These were the `ExecutionPlanVisitor`s I made.  I would be happy to add docs around these.

```
struct FileScanVisitor {
        file_scan_config: Option<FileScanConfig>,
    }

    impl ExecutionPlanVisitor for FileScanVisitor {
        type Error = anyhow::Error;

        fn pre_visit(&mut self, plan: &dyn ExecutionPlan) -> Result<bool, Self::Error> {
            let maybe_parquet_exec = plan.as_any().downcast_ref::<ParquetExec>();
            if let Some(parquet_exec) = maybe_parquet_exec {
                self.file_scan_config = Some(parquet_exec.base_config().clone());
            }
            Ok(true)
        }
    }

    fn get_file_scan_config(plan: Arc<dyn ExecutionPlan>) -> Option<FileScanConfig> {
        let mut visitor = FileScanVisitor {
            file_scan_config: None,
        };
        visit_execution_plan(plan.as_ref(), &mut visitor).unwrap();
        visitor.file_scan_config
    }
```

```
#[derive(Debug)]
struct ParquetVisitor;

impl ExecutionPlanVisitor for ParquetVisitor {
    type Error = DataFusionError;

    fn pre_visit(&mut self, plan: &dyn ExecutionPlan) -> Result<bool, Self::Error> {
        // Get the one-line representation of the ExecutionPlan, something like this:
        //   ParquetExec: file_groups=[...], ...
        let mut buf = String::new();
        write!(&mut buf, "{}", displayable(plan).one_line()).map_err(|e| {
            DataFusionError::Internal(format!("Error while collecting metrics: {e}"))
        })?;

        // Trim everything up to the first colon.
        // This is a hack to extract a human-readable representation of the ExecutionPlan's type.
        // We would prefer if `ExecutionPlan` had `name` method, but this will do,
        // since every physical operator seems to follow this convention.
        // If a node doesn't, we just skip collecting its metrics, and no harm is done.
        let plan_type = match buf.split_once(':') {
            None => {
                println!("execution plan has unexpected display format: {buf}");
                return Ok(true);
            }
            Some((name, _)) => name.to_string(),
        };
        let maybe_parquet_exec = plan.as_any().downcast_ref::<ParquetExec>();
        match maybe_parquet_exec {
            Some(parquet_exec) => {
                let metrics = match parquet_exec.metrics() {
                    None => return Ok(true),
                    Some(metrics) => metrics,
                };
                // println!("Metrics: {:?}", metrics);
                let bytes_scanned = metrics.sum_by_name("bytes_scanned");
                println!("Parquet Bytes scanned: {:?}", bytes_scanned);
            }
            None => {

            }
        }
        Ok(true)
    }
}
```

I had in mind having two examples, one for getting information from parquet files (I could probably combine the two I had) and one that tracked data across all nodes (maybe `output_rows`).

### PR DESCRIPTION
**Pull request by matthewmturner**
## Which issue does this PR close?

Closes #10012

### CODE REVIEW

**Comment by matthewmturner** [comment]
@alamb FYI

File: datafusion-examples/examples/parquet_exec_visitor.rs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    // For example, in this example the `file_scan_config` is known without executing
+    // but the `bytes_scanned` would be None if we did not execute.
+    let mut batch_stream = execute_stream(plan.clone(), ctx.task_ctx()).unwrap();
+    while let Some(batch) = batch_stream.next().await {
+        println!("Do something with batch");
+    }
+
+    visit_execution_plan(plan.as_ref(), &mut visitor).unwrap();
+
+    println!("ParquetExecVisitor: {:?}", visitor);
```
Comment (alamb) [inline datafusion-examples/examples/parquet_exec_visitor.rs:74]: The output looks a little messy:

```shell
Do something with batch
ParquetExecVisitor: ParquetExecVisitor { file_scan_config: Some(object_store_url=ObjectStoreUrl { url: Url { scheme: "file", cannot_be_a_base: false, username: "", password: None, host: None, port: None, path: "/", query: None, fragment: None } }, statistics=Statistics { num_rows: Exact(8), total_byte_size: Exact(671), column_statistics: [ColumnStatistics { null_count: Absent, max_value: Absent, min_value: Absent, distinct_count: Absent }, ColumnStatistics { null_count: Absent, max_value: Absent, min_value: Absent, distinct_count: Absent }, ColumnStatistics { null_count: Absent, max_value: Absent, min_value: Absent, distinct_count: Absent }, ColumnStatistics { null_count: Absent, max_value: Absent, min_value: Absent, distinct_count: Absent }, ColumnStatistics { null_count: Absent, max_value: Absent, min_value: Absent, distinct_count: Absent }, ColumnStatistics { null_count: Absent, max_value: Absent, min_value: Absent, distinct_count: Absent }, ColumnStatistics { null_count: Absent, max_value: Absent, min_value: Absent, distinct_count: Absent }, ColumnStatistics { null_count: Absent, max_value: Absent, min_value: Absent, distinct_count: Absent }, ColumnStatistics { null_count: Absent, max_value: Absent, min_value: Absent, distinct_count: Absent }, ColumnStatistics { null_count: Absent, max_value: Absent, min_value: Absent, distinct_count: Absent }, ColumnStatistics { null_count: Absent, max_value: Absent, min_value: Absent, distinct_count: Absent }] }, file_groups={1 group: [[Users/andrewlamb/Software/datafusion2/parquet-testing/data/alltypes_plain.parquet]]}, projection=[id, bool_col, tinyint_col, smallint_col, int_col, bigint_col, float_col, double_col, date_string_col, string_col, timestamp_col]), bytes_scanned: Some(Count { name: "bytes_scanned", count: Count { value: 671 } }) }
```

Maybe we could print out a nicer version like 

```rust

    let bytes_scanned = visitor.bytes_scanned.unwrap();
    println!("Total parquet bytes scanned {bytes_scanned}");
```

```shell
Do something with batch
Total parquet bytes scanned 671
```

File: datafusion-examples/examples/parquet_exec_visitor.rs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+#[derive(Debug)]
+struct ParquetExecVisitor {
+    file_scan_config: Option<FileScanConfig>,
+    bytes_scanned: Option<MetricValue>,
+}
+
+impl ExecutionPlanVisitor for ParquetExecVisitor {
+    type Error = datafusion_common::DataFusionError;
+
+    /// Based on your needs implement either `pre_visit` or `post_visit`
```
Comment (alamb) [inline datafusion-examples/examples/parquet_exec_visitor.rs:87]:
```suggestion
    /// This function is called once for every node in the tree. 
    /// Based on your needs implement either `pre_visit` (visit each node before its children/inputs)
    /// or `post_visit` (visit each node after its children/inputs)
```

**Review by alamb — APPROVED** [review]
Thank you @matthewmturner  -- this looks great and I think will really help. 

It looks like clippy is complaining in CI: https://github.com/apache/datafusion/actions/runs/8878883209/job/24375602460?pr=10286 (maybe you could print the batch count)

Once the CI is fixed I think we can merge this PR, though I had a few suggestions for improvmements: 

1. I left some suggestions for additional comments for context
2. I think we should also add an entry to the example readme [explaining](https://github.com/apache/datafusion/tree/main/datafusion-examples#readme) what this shows (e.g. "Extract statistics by visiting an `ExecutionPlan`  after execution" or something)

**Comment by matthewmturner** [comment]
Thanks @alamb, I have updated.

**Review by alamb — APPROVED** [review]
Thank you @matthewmturner  -- looks very helpful to me

File: datafusion-examples/examples/parquet_exec_visitor.rs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        "ParquetExecVisitor bytes_scanned: {:?}",
+        visitor.bytes_scanned
+    );
+    println!(
+        "ParquetExecVisitor file_groups: {:?}",
+        visitor.file_scan_config.unwrap().file_groups
+    );
+}
+
+/// Define a struct with fields to hold the information you want to collect
```
Comment (comphead) [inline datafusion-examples/examples/parquet_exec_visitor.rs:81]: This comment is very chatgpt-like :))

Comment (matthewmturner) [reply datafusion-examples/examples/parquet_exec_visitor.rs:81]: Aha I promise I did not use ChatGPT to make this.

Maybe need to add fine print like "LLMs were not used in the generation of this code" 😆

Comment (comphead) [reply datafusion-examples/examples/parquet_exec_visitor.rs:81]: :) Lets refine struct doc to be more specific

Comment (matthewmturner) [reply datafusion-examples/examples/parquet_exec_visitor.rs]: I have updated

**Review by comphead — APPROVED** [review]
lgtm @matthewmturner thanks for addressing the comments. Im planning to merge it once CI completed

**Pull request merged**